Refactor/solve continue watching issues

Noor-Serry opened the pull request:
# Pull Request Template

## Description
in this branch I do the solve the following issues for continue watching screen :
- Continue watching doesn't update when the device language change
- Sort Continue Watching By Last Time Watched For Mixed Tv and Movie
- add pagination to continue watching
- Continue watching repeat the same item twice, one for Arabic language and other for English (if i change the device language)
---

## Changes Made
List the changes introduced in this pull request:

- MovieWatchHistory && TvShowWatchHistory  I make those entity to have movie , tvshow entity insted of the id to make the sort easiest since each entity have the lastWatchedTime and the movie or tv show that will diplayed 
- I modified the watch history repository and the watch history data source for tv and movie now the follow is when the user change the device languish I get the watchied page as ids then loop in those ids and try to get the movie from local if it return null then I will fetch it from the api then cache it then return it from the local same thing happen with the tvshow  

---

## Screenshots (if applicable)
Include screenshots or GIFs to showcase the changes, especially if they impact the UI.
[![Watch the demo](https://img.youtube.com/vi/yOoEMIeLmeQ/0.jpg)](https://youtu.be/yOoEMIeLmeQ)


---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

---

## Additional Comments
Add any additional information or context about the pull request here.

Commit:
feature: implement paging for continue watching items

This commit introduces paging functionality for the "Continue Watching" section in both the dedicated Continue Watching screen and the home screen.

Key changes include:

- **Refactored Data Fetching:**
    - `GetContinueWatchingScreenDataUseCase` now accepts `page` and `pageSize` parameters.
    - `WatchHistoryLocalDataSource` methods for fetching continue watching movies and TV shows now support pagination (`offset`, `limit`).
    - SQL queries in `WatchHistoryDao` are updated to include `LIMIT` and `OFFSET` clauses.
- **Updated Entities and DTOs:**
    - `MovieWatchHistory` and `TvShowWatchHistory` entities now hold the respective `Movie` and `TvShow` objects instead of just IDs.
    - `MovieWatchHistoryDto` and `TvShowWatchHistoryDto` no longer store `lastWatchedTime`; this information is now derived from the `insertedDate` of `LocalMovieDto` and `LocalTvShowDto`.
    - `LocalMovieDto` and `LocalTvShowDto` now include an `insertedDate` field.
- **New Mappers:**
    - Introduced `MovieWatchHistoryLocalMapper` and `TvWatchHistoryLocalMapper` to map `LocalMovieDto`/`LocalTvShowDto` to their respective watch history entities.
    - Removed `MovieWatchHistoryMapper` and `TvShowWatchHistoryMapper`.
- **ViewModel and UI Updates:**
    - `ContinueWatchingViewModel` now uses `Pager` to load data in a paginated manner.
    - `ContinueWatchingUiState`'s `continueMediaItemUiStates` is now a `Flow<PagingData<ContinueWatchingMediaItemUiState>>`.
    - `ContinueWatchingScreen` and `ContinueWatchingMediaItemsGrid` are updated to work with `LazyPagingItems`.
    - `HomeViewModel` is updated to fetch and display paginated continue watching items on the home screen.
    - `HomeUiState` and `ContinueWatchingUiStateMapper` now use `ContinueWatchingMediaItemUiState` which includes `dateAdded` for sorting.
- **Domain Layer Adjustments:**
    - Use cases for adding watch history (`AddMovieWatchHistoryUseCase`, `AddTvShowWatchHistoryUseCase`) now only take the media ID.
    - Use cases for getting continue watching items (`GetContinueWatchingMoviesUseCase`, `GetContinueWatchingTvShowsUseCase`) now accept pagination parameters and return `Flow<List<MovieWatchHistory>>` and `Flow<List<TvShowWatchHistory>>` respectively.
- **Repository Layer Adjustments:**
    - `WatchHistoryRepository` methods are updated to align with domain layer changes and now pass pagination parameters to the local data source.

Commit:
feature: implement paging for continue watching items

This commit introduces paging functionality for the "Continue Watching" section in both the dedicated Continue Watching screen and the home screen.

Key changes include:

- **Refactored Data Fetching:**
    - `GetContinueWatchingScreenDataUseCase` now accepts `page` and `pageSize` parameters.
    - `WatchHistoryLocalDataSource` methods for fetching continue watching movies and TV shows now support pagination (`offset`, `limit`).
    - SQL queries in `WatchHistoryDao` are updated to include `LIMIT` and `OFFSET` clauses.
- **Updated Entities and DTOs:**
    - `MovieWatchHistory` and `TvShowWatchHistory` entities now hold the respective `Movie` and `TvShow` objects instead of just IDs.
    - `MovieWatchHistoryDto` and `TvShowWatchHistoryDto` no longer store `lastWatchedTime`; this information is now derived from the `insertedDate` of `LocalMovieDto` and `LocalTvShowDto`.
    - `LocalMovieDto` and `LocalTvShowDto` now include an `insertedDate` field.
- **New Mappers:**
    - Introduced `MovieWatchHistoryLocalMapper` and `TvWatchHistoryLocalMapper` to map `LocalMovieDto`/`LocalTvShowDto` to their respective watch history entities.
    - Removed `MovieWatchHistoryMapper` and `TvShowWatchHistoryMapper`.
- **ViewModel and UI Updates:**
    - `ContinueWatchingViewModel` now uses `Pager` to load data in a paginated manner.
    - `ContinueWatchingUiState`'s `continueMediaItemUiStates` is now a `Flow<PagingData<ContinueWatchingMediaItemUiState>>`.
    - `ContinueWatchingScreen` and `ContinueWatchingMediaItemsGrid` are updated to work with `LazyPagingItems`.
    - `HomeViewModel` is updated to fetch and display paginated continue watching items on the home screen.
    - `HomeUiState` and `ContinueWatchingUiStateMapper` now use `ContinueWatchingMediaItemUiState` which includes `dateAdded` for sorting.
- **Domain Layer Adjustments:**
    - Use cases for adding watch history (`AddMovieWatchHistoryUseCase`, `AddTvShowWatchHistoryUseCase`) now only take the media ID.
    - Use cases for getting continue watching items (`GetContinueWatchingMoviesUseCase`, `GetContinueWatchingTvShowsUseCase`) now accept pagination parameters and return `Flow<List<MovieWatchHistory>>` and `Flow<List<TvShowWatchHistory>>` respectively.
- **Repository Layer Adjustments:**
    - `WatchHistoryRepository` methods are updated to align with domain layer changes and now pass pagination parameters to the local data source.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into refactor/add-paging-to-continue-watching

# Conflicts:
#	ui/build.gradle.kts
#	ui/src/main/java/com/amsterdam/ui/screens/continueWatching/component/ContinueWatchingMediaItemsGrid.kt

Commit:
refactor: Group kotlinx-datetime dependency

This commit groups the `kotlinx-datetime` dependency into a dedicated `kotlinxDateTime()` function within the `ui/build.gradle.kts` file for better organization and readability.

Commit:
refactor: Update page size for continue watching pagination

The page size for the `Pager` in `ContinueWatchingViewModel` has been increased from 8 to 20. This change applies to both the `PagingConfig` and the `pageSize` parameter passed to `getContinueWatchingScreenDataUseCase`.

Commit:
refactor: improve watch history and multi-language support

This commit introduces several key changes:

- **Refactors Watch History:**
    - `WatchHistoryLocalDataSource` now returns `MovieWatchHistoryDto` and `TvShowWatchHistoryDto` directly instead of `LocalMovieDto` and `LocalTvShowDto`.
    - `WatchHistoryRepositoryImpl` is updated to handle this change. It now fetches movie/TV show details from local or remote sources if not already cached when retrieving watch history.
    - `MovieWatchHistoryDto` and `TvShowWatchHistoryDto` no longer store `storedLanguage`, as this is now handled by the repository.
- **Enhances Multi-language Support:**
    - `MovieDao`, `TvShowDao`, `MovieLocalSource`, and `TvShowLocalSource` methods for fetching by ID now require a `storedLanguage` parameter.
    - `HomeScreen`, `HomeViewModel`, `ContinueWatchingScreen`, and `ContinueWatchingViewModel` now observe locale changes and reload data accordingly to reflect the current device language.
- **Removes ServiceProvider Layer:**
    - Deletes `MovieServiceProvider`, `MovieServiceProviderImpl`, `TvShowsServiceProvider`, and `TvShowsServiceProviderImpl` as this abstraction layer was deemed unnecessary. Remote data source interactions will now be handled more directly.

Commit:
refactor: extract methods for fetching watch history

This commit refactors the `WatchHistoryRepositoryImpl` by extracting the logic for fetching movie and TV show watch history into separate private methods: `getMovieWatchHistory` and `getTvShowWatchHistory`.

This improves code readability and maintainability by encapsulating the fetching and caching logic for each media type. The main `getMoviesWatchHistory` and `getTvShowsWatchHistory` methods now simply collect from the local data source and map the results using these new helper functions.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into refactor/add-paging-to-continue-watching

# Conflicts:
#	ui/src/main/java/com/amsterdam/ui/screens/home/HomeScreen.kt

Commit:
refactor: remove redundant import

The import `androidx.compose.ui.platform.LocalConfiguration` was present twice in `HomeScreen.kt`. This commit removes the duplicate import.

Commit:
fix: remove unnecessary rate formatting in ContinueWatchingSection

The `formateAsRate()` extension function was incorrectly used on the `item.rate` property within the `ContinueWatchingSection`. This function is intended for formatting numerical ratings, while `item.rate` already represents a formatted string (e.g., "PG-13").

This commit removes the redundant call to `formateAsRate()`, ensuring the movie rating is displayed correctly.

saraashraf11 commented on entity/src/main/java/com/amsterdam/entity/MovieWatchHistory.kt:
```diff
@@ -3,6 +3,6 @@ package com.amsterdam.entity
 
 data class MovieWatchHistory(
-    val movieId: Long,
+    val movie: Movie,
```
Why did you change the name from movieId to movie? I think it was clearer in the beginning, or what do you think?

Noor-Serry replied:
I change the variable type to Movie so that I changed the variable name , 
why I did that to remove redundant complexly that will be added if it was id

saraashraf11 commented on entity/src/main/java/com/amsterdam/entity/TvShowWatchHistory.kt:
```diff
@@ -3,6 +3,6 @@ package com.amsterdam.entity
 
 data class TvShowWatchHistory(
-    val tvShowId: Long,
+    val tvShow: TvShow,
```
Also here

Noor-Serry replied:
same as above

nabilBouzineDev commented on repository/src/main/java/com/amsterdam/repository/repository/WatchHistoryRepositoryImpl.kt:
```diff
@@ -1,54 +1,115 @@
+    override fun getContinueWatchingMovies(
+        page: Int,
+        pageSize: Int
+    ): Flow<List<MovieWatchHistory>> = flow {
```
@Noor-Serry I don't get it, why you are collecting inside a flow and emit it at the same time?

Noor-Serry replied:
that is what I do in this part :
I get the watchied page as ids then loop in those ids and try to get the movie from local if it return null then I will fetch it from the api then cache it then return it from the local same thing happen with the tvshow

nabilBouzineDev replied:
But, you are collecting from the local data source already why you need to double check it here as well
`movieLocalSource.getMovieById(movieWatchHistoryDto.movieId, getDeviceLanguage())` in line 53

Noor-Serry replied:
I did not got you clearly but here as said get the movie from local if its return null then make api call then cache then return it from local

nabilBouzineDev commented on viewModel/src/main/java/com/amsterdam/viewmodel/continueWatching/ContinueWatchingViewModel.kt:
```diff
@@ -84,5 +98,12 @@ class ContinueWatchingViewModel @Inject constructor(
         sendNewEffect(ContinueWatchingEffect.NavigateBack)
     }
 
+    fun onLanguageChanged(newLanguish: Locale) {
```
There's a typo here and also please wait until this PR is merged #293 to handle this local changes.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into refactor/solve-continue-watching-issues

# Conflicts:
#	repository/src/main/java/com/amsterdam/repository/mapper/local/MovieWatchHistoryMapper.kt
#	repository/src/main/java/com/amsterdam/repository/mapper/local/TvShowWatchHistoryMapper.kt
#	repository/src/main/java/com/amsterdam/repository/repository/WatchHistoryRepositoryImpl.kt
#	ui/src/main/java/com/amsterdam/ui/screens/continueWatching/ContinueWatchingScreen.kt
#	ui/src/main/java/com/amsterdam/ui/screens/home/HomeScreen.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/continueWatching/ContinueWatchingViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/home/HomeViewModel.kt

Commit:
feat: make continue watching section update on language change

This commit updates the `WatchHistoryRepositoryImpl` to use `AppPreferences` for getting the device language.
It also modifies `ContinueWatchingViewModel` to handle language changes and refresh the continue watching data accordingly.
The `ContinueWatchingScreen` no longer needs to observe locale changes directly, as the ViewModel now handles this.
Additionally, `HomeViewModel` is updated to re-fetch continue watching data when the language changes.

Commit:
feature: store watched date in watch history

This commit adds a `watchedDate` field to `MovieWatchHistoryDto` and `TvShowWatchHistoryDto` to store the timestamp when a movie or TV show was watched.

The `WatchHistoryRepositoryImpl` is updated to use this `watchedDate` when mapping local DTOs to entities.

Additionally, this commit includes:
- Logging of page numbers in `PagingSource`.
- A fix in `PopularSection.kt` to prevent rendering if `mediaItems` is empty.
- Removal of unused `LocalConfiguration` import in `ContinueWatchingScreen.kt`.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into refactor/solve-continue-watching-issues

# Conflicts:
#	repository/src/main/java/com/amsterdam/repository/repository/WatchHistoryRepositoryImpl.kt
